### PR TITLE
feat: KeyStreams

### PR DESCRIPTION
**Pull request by xavierchanth**
**- What I did**

Add an abstraction layer around the notificationService subscription which allows the app developer to observe a set of keys as a Stream. There are various types included:

- Single Element Stream (any generic non-collection type) [KeyStream\<T>]
- Iterable [IterableKeyStream\<T>]
- List [ListKeyStream\<T>]
- Set [SetKeyStream\<T>]
- Map [MapKeyStream<K,V>]

**- How I did it**

Local changes and e2e testing

**- How to verify it**

Read the mixin then two base files first when reviewing.
See the tests I wrote

**- Description for the changelog**
Feat: KeyStreams

### CODE REVIEW

**Comment by xavierchanth** [comment]
@gkc this feature should be good to go now. I used templates/macros to handle the dart docs, do you know of any better solutions?

**Comment by gkc** [comment]
Hi @XavierChanth - will take me a while to review this as there's a lot here, so I may not get to it until Tuesday (but will try to do sooner if I can)

**Comment by xavierchanth** [comment]
> Hi @XavierChanth - will take me a while to review this as there's a lot here, so I may not get to it until Tuesday (but will try to do sooner if I can)

No worries! It is quite a lot, I tried my best to make the implementation as simple as possible. There are a lot of similar patterns which should make reading easier.

**Review by gkc — CHANGES_REQUESTED** [review]
Very nice! LGTM in general

I have a few requests
1. Need tests verifying that this all works as intended in the event of switching between atsigns
2. Some documentation targeted at package users (maybe in key_stream.dart)  describing typical usages
3. Maybe some additional documentation clarifying or explaining behaviours when switching between atsigna
4. Some documentation targeted at package maintainers (in key_stream_mixin.dart) explaining how it hooks in with notifications

**Comment by xavierchanth** [comment]
> Very nice! LGTM in general
> 
> I have a few requests
> 
> 1. Need tests verifying that this all works as intended in the event of switching between atsigns
> 2. Some documentation targeted at package users (maybe in key_stream.dart)  describing typical usages
> 3. Maybe some additional documentation clarifying or explaining behaviours when switching between atsigna
> 4. Some documentation targeted at package maintainers (in key_stream_mixin.dart) explaining how it hooks in with notifications

Noted, will work on these when I get the chance.

File: at_client/lib/src/key_stream/key_stream_mixin.dart [hunk 1]
```diff
@@ -0,0 +1,419 @@
...
+  /// A subscription that listens to notifications.
+  @visibleForTesting
+  late StreamSubscription<AtNotification> notificationSubscription;
+
+  late AtClientManager _atClientManager;
+
+  static final AtSignLogger _logger = AtSignLogger('KeyStream');
+
+  @visibleForTesting
+  static bool disposeOnAtsignChange = true;
```
Comment (xavierchanth) [inline at_client/lib/src/key_stream/key_stream_mixin.dart]: @gkc there was in fact was an instance where we switch atsigns. I've added this flag to check whether we should listen to the atsign change.

Comment (gkc) [reply at_client/lib/src/key_stream/key_stream_mixin.dart]: 👍👍👍

**Comment by xavierchanth** [comment]
@gkc let me know what you think, I've updated the documentation.

**Review by gkc — APPROVED** [review]
LGTM - really good work.

**Comment by xavierchanth** [comment]
@gkc requesting another review for:

- Added onError function to catch exceptions from `atClient.get` calls.
- For each collection, removed `<collection>_key_stream_impl.dart` and merged with `<collection>_key_stream.dart`.
    (i.e. removed an extra and unnecessary layer)

**Review by gkc — APPROVED** [review]
LGTM. There is a functional test failing; although it seems unrelated, plz take a look

**Comment by gkc** [comment]
@kalluriramkumar was the failing test here due to the same issue which you uncovered?

**Comment by xavierchanth** [comment]
> @kalluriramkumar was the failing test here due to the same issue which you uncovered?

@kalluriramkumar if it is, please merge.

**Pull request merged**